server/src/main/java/keywhiz/service/resources/admin/SecretsResource.java

```diff
@@ -487,6 +490,135 @@ public Response deleteSecret(@Auth User user, @PathParam("secretId") LongParam s
+    Optional<Secret> secret = secretController.getSecretByName(secretName);
+    if (secret.isPresent()) {
+      logger.info("User '{}' tried renaming a secret, but another secret with that name "
+              + "already exists (name={})", user, secretId.get());
+      throw new NotAllowedException("That name is already taken by another secret"); //use different error
+    }
+
+    logger.info("User '{}' renamed secret id={} to name='{}'", user, secretId,
+        secretName);
+
+    secretDAOReadWrite.renameSecretById(secretId.get(), secretName);
+
+    // Record the rename
+    Map<String, String> extraInfo = new HashMap<>();
+    auditLog.recordEvent(
+        new Event(Instant.now(), EventTag.SECRET_RENAME, user.getName(), secretName,
+            extraInfo));
+    return Response.noContent().build();
+  }
+
+  /**
+   * Retrieve all Secrets Contents associated with the given Secret ID
+   *
+   * @param user     the admin user performing this operation
+   * @param secretId the ID of the Secret that will be associated with the secrets contents returned
+   * @return 200 if secret deleted, 404 if not found
+   * <p>
+   * description Returns a list of secrets contents for a given Secret ID.
+   * <p>
+   * responseMessage 200 Found and deleted Secret with given ID
+   * <p>
+   * responseMessage 404 Secret with given ID not Found
+   */
+  @Path("secretcontents/list/{secretId}")
+  @Timed @ExceptionMetered
+  @POST
+  public List<SecretSeriesAndContent> getSecretsContentsBySecretId(@Auth User user,
+      @PathParam("secretId") LongParam secretId, @QueryParam("idx") Integer idx,
+      @QueryParam("num") Integer num) {
+
+    ImmutableList<SecretSeriesAndContent> secretsContents =
+        secretDAOReadOnly.getDeletedSecretVersionsBySecretId(secretId.get(), idx, num)
+            .orElseThrow(NotFoundException::new);
+
+    logger.info("User '{}' retrieving secrets contents for secretId={} from idx={} for {} secrets contents.",
+        user, secretId, idx, num);
+
+    return secretsContents;
+  }
+
+  /**
+   * Update Secret content for the given secret (identified by secretId) to secretsContentId
+   *
+   * @param user     the admin user performing this operation
+   * @param secretId the ID of the Secret to be updated
+   * @param secretsContentId, the Secret content to be updated
+   * @return 200 if secret deleted, 404 if not found
+   * <p>
+   * description Updates a secret current for a given secret. Used by Keywhiz CLI and the web ui.
+   * <p>
+   * responseMessage 200 Found and deleted Secret with given ID
+   * <p>
+   * responseMessage 404 Secret with given ID not Found
+   */
+  @Path("secretcontents/update/{secretsContentId}/{secretId}")
+  @Timed @ExceptionMetered
+  @POST
+  public Response updateSecretsCurrent(@Auth User user, @PathParam("secretId") LongParam secretId,
+      @PathParam("secretsContentId") LongParam secretsContentId) {
+    Optional<Secret> secret = secretController.getSecretById(secretId.get());
```

Comment:
Is the plan for this endpoint to be used to restore deleted secrets? If so, we should actually remove this check because `getSecretById` works only for non-deleted secrets.